net: disconnect peers by address without using a subnet

vasild opened the pull request:
Previously, when disconnecting a peer with a given address we would
create a dummy subnet that contains just 1 address (/32 for IPv4 and
/128 for IPv6) and would disconnect all peers that belong to this
subnet (there may be more than one connection, to a different port).

The problem is that for any non-IPv4 and non-IPv6 address we would
create an invalid subnet which would later not match any addresses.

Thus, don't use a one-host-subnet, but compare the addresses directly.
This works for any type of addresses.

<!--
*** Please remove the following help text before submitting: ***

Pull requests without a rationale and clear improvement may be closed
immediately.

GUI-related pull requests should be opened against
https://github.com/bitcoin-core/gui
first. See CONTRIBUTING.md
-->

<!--
Please provide clear motivation for your patch and explain how it improves
Bitcoin Core user experience or Bitcoin Core developer experience
significantly:

* Any test improvements or new tests that improve coverage are always welcome.
* All other changes should have accompanying unit tests (see `src/test/`) or
  functional tests (see `test/`). Contributors should note which tests cover
  modified code. If no tests exist for a region of modified code, new tests
  should accompany the change.
* Bug fixes are most welcome when they come with steps to reproduce or an
  explanation of the potential issue as well as reasoning for the way the bug
  was fixed.
* Features are welcome, but might be rejected due to design or scope issues.
  If a feature is based on a lot of dependencies, contributors should first
  consider building the system outside of Bitcoin Core, if possible.
* Refactoring changes are only accepted if they are required for a feature or
  bug fix or otherwise improve developer experience significantly. For example,
  most "code style" refactoring changes require a thorough explanation why they
  are useful, what downsides they have and why they *significantly* improve
  developer experience or avoid serious programming bugs. Note that code style
  is often a subjective matter. Unless they are explicitly mentioned to be
  preferred in the [developer notes](/doc/developer-notes.md), stylistic code
  changes are usually rejected.
-->

<!--
Bitcoin Core has a thorough review process and even the most trivial change
needs to pass a lot of eyes and requires non-zero or even substantial time
effort to review. There is a huge lack of active reviewers on the project, so
patches often sit for a long time.
-->

vasild committed:
net: disconnect peers by address without using a subnet

Previously, when disconnecting a peer with a given address we would
create a dummy subnet that contains just 1 address (/32 for IPv4 and
/128 for IPv6) and would disconnect all peers that belong to this
subnet (there may be more than one connection, to a different port).

The problem is that for any non-IPv4 and non-IPv6 address we would
create an invalid subnet which would later not match any addresses.

Thus, don't use a one-host-subnet, but compare the addresses directly.
This works for any type of addresses.

vasild commented:
The problem was discovered by @MarcoFalke.

MarcoFalke commented:
tested-only ACK 5fd2cee7fbc38c677d08a1545770156cc061833e did not review the code 🔮

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

tested-only ACK 5fd2cee7fbc38c677d08a1545770156cc061833e did not review the code 🔮
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUh/Sgv/W9d+TPMjSm+JqcwJBW7n3wmto0PDu+CvGkHSwHJaoLU7toj8W1y6Ohfz
CDKRGq2ZG+VnxSbKRhItSeaWqleHfAJ0Ns1xZNbDudAw7/IrLMsvNd7+/apFgpUT
iM1ytrZMRRLqoPdRVgpzkamKMUyPTPzzXzGQoH6xdxSjMuKbJ69DVzg4d0DuAmH+
y+ZP85yi7KDq5YvS25JUaOpZg4m9UnwTHctL1+hMeIWBtxlx3pFohExswbUC28xz
ytN4dWFxyQXC/t3i3fY3AiLiHZZ0frxHTd26sL/vqj935MWnH3SsIpIwm57kl1v2
vHR7yakgWEcDF5qEQHa1+wYi1p26BgYQYjlPqYCM4fp+0MBg6lPpggthZkYVKTXI
Jg2mxdUl/kuFDA8Xy9u00uQj6BOWJfshBuClQkxf5184qTcMdQ6SwxagpfDGHEjn
JDHbX+djj8+CfR6CVc3P7MWP1UxnuFI8NWiyMI6KrcEH3ESw16LmcrS3KOeLLI4Q
5IVN+71Z
=kXbR
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `ffb08779a664f7e34db6952de6f0948d1fb8eeb7089b81525f9ffde67a8ac202  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108ffb08779a664f7e34db6952de6f0948d1fb8eeb7089b81525f9ffde67a8ac202f010ba914ef96fcd58dcf7ef97e69507f30f08fff01001ca557449cdd59ee8a489fa048bb29a08f1045ff343ccf0088c83c2e0c2202b560083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff01004a917a47de2960e291f72db62a88de908f020750bb0dd61da48daa72731e0113a36703767a6b3fa7aedbd93d981a8f0b83d2008f020552fdc0c24dad87b2a713d4c04937ee0f5bbcd61bb355994634f12a8e8c28cdc08f1045ff343ccf00861a940fef48bc4080083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff01016190e20a52f4e363ae04d98e1a52d8a08f020cd808e86f44edce43d5e8b4acfb96b5f61594caba830320a4d1be42eb59ebf1708f02022722abb7230b220a5da7037cd78b817045cf688c776916a9e38b955dc9a124608f02050b970d8126c0f5156e36ddb8b6f7d12297ee07c6860729af27d60d4df312ed908f1045ff343ccf0088f329964afa83b3d0083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010dce13a9487c5c580d62df60b7705cc0d08f020b00c82767aafbe20855a385ea3e94bc9d48475c132a5000324b368a74520fba808f020ce0e21933b7b4cf9b6f557430dc02455a94ae6c0e85cee26ded22d0094a9dad408f1045ff343ccf00850484396068324400083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

jonatack commented:
If this is a bug, it dates back to 7cc2b9f6786 at least.

jonatack reviewed (commented):
Code review ACK  5fd2cee7fbc38c677d08a1545770156cc061833e

It would be good to have a failing test case that this patch fixes, either unit tests or in the functional `rpc_setban` or `p2p_disconnect_ban` ones.

jonatack commented:
Adding some logging like this to check before/after outbound onion disconnection, grepping the log for "subnet"
```cpp
LogPrintf("Disconnecting subnet peer=%d address=%s\n network=%s", pnode->GetId(),
          pnode->addr.ToString(), GetNetworkName(pnode->ConnectedThroughNetwork()));
```

vasild commented:
https://github.com/bitcoin/bitcoin/pull/20852 contains a fix for this issue + it also fixes a similar issue in `BanMan`. If it is merged, then this can be closed without merge. They do not conflict in any way. Both can go in.

jonatack commented:
Any tips how to test this manually? Disconnecting outbound onion peers with `disconnectnode` works for me on both master and this PR and in #20852.

vasild commented:
@jonatack yes, the `disconnectnode` RPC uses either `DisconnectNode(std::string)` or `DisconnectNode(NodeId)` and the problem is in `DisconnectNode(CNetAddr)`.

`bitcoin-cli setban foo.onion add` will do the job.

taurus228 commented:
вт, 5 янв. 2021 г. в 19:24, Vasil Dimov <notifications@github.com>:

> @jonatack <https://github.com/jonatack> yes, the disconnectnode RPC uses
> either DisconnectNode(std::string) or DisconnectNode(NodeId) and the
> problem is in DisconnectNode(CNetAddr).
>
> bitcoin-cli setban foo.onion add will do the job.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/bitcoin/bitcoin/pull/20849#issuecomment-754740638>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AK4ZHQKMTWFQ66BD3TC7WTLSYM4J7ANCNFSM4VTHTMFA>
> .
>
-- 
Nik

sipa commented:
utACK 5fd2cee7fbc38c677d08a1545770156cc061833e

Let's get this merged so the path forward for 0.21 is clear (we can still do #20852, but it's no longer critical then).

jonatack commented:
Adding logging to the code path and verified that `setban <onion> add` disconnects peers with this patch and fails to do so on master @ 68196a891056f, with both tor v2 and v3 peers.

Logging added to master and to this patch:
```diff
@@ -2805,6 +2805,7 @@ bool CConnman::DisconnectNode(const CSubNet& subnet)
     LOCK(cs_vNodes);
     for (CNode* pnode : vNodes) {
         if (subnet.Match(pnode->addr)) {
+            LogPrintf("Disconnecting %s via subnet peer=%d address=%s network=%s\n",
+                      pnode->ConnectionTypeAsString(), pnode->GetId(),
+                      pnode->addr.ToString(), GetNetworkName(pnode->ConnectedThroughNetwork()));
             pnode->fDisconnect = true;
             disconnected = true;
         }
```
Logging also added to this patch:
```diff
@@ -2818,6 +2819,7 @@ bool CConnman::DisconnectNode(const CNetAddr& addr)
     LOCK(cs_vNodes);
     for (CNode* pnode : vNodes) {
         if (addr == pnode->addr) {
+            LogPrintf("Disconnecting %s via addr peer=%d address=%s network=%s\n",
+                      pnode->ConnectionTypeAsString(), pnode->GetId(),
+                      pnode->addr.ToString(), GetNetworkName(pnode->ConnectedThroughNetwork()));
             pnode->fDisconnect = true;
             disconnected = true;
         }
```

On master, disconnection fails and nothing is logged. With this patch:
```
2021-01-06T11:47:36Z Disconnecting manual via addr peer=5 address=TOR_V3_PEER network=onion
2021-01-06T11:54:30Z Disconnecting manual via addr peer=25 address=TOR_V2_PEER network=onion
2021-01-06T11:54:49Z Disconnecting outbound-full-relay via addr peer=16 address=TOR_V2_PEER:8333 network=onion
```

Upgrading to tested ACK 5fd2cee7fbc38c677d08a1545770156cc061833e

vasild committed:
test: add test to ensure non-IP peers are properly misbehaved

Change the `peer_discouragement` test to use `CNode` pointers so that
the nodes it uses can be added to `CConnman::vNodes` and cleaned up
properly. Make it use `CConnmanTest` instead of `CConnman`. This is
needed because we want to check `CNode::fDisconnect` and for this flag
to be flipped by `CConnman::DisconnectNode()` the node must be in
`CConnman::vNodes`.

Extend the test with one torv3 peer and check that it is discouraged and
disconnected as expected.

vasild commented:
Appended an extra commit with unit test that fails on `master` and passes on this PR.

If it does not get re-ACKs or if it turns out to be problematic I will remove it, restoring the already ACK'ed commit id.

jonatack commented:
Nice, will review asap.

jonatack reviewed (commented):
ACK 98140bc855b037d2c552153496e8e4e2c2efdad9 code review, manual testing with `setban OUTBOUND_ONION add`, and verified the new test fails on master on the non-IP Tor v3 peer disconnection assertion.

laanwj commented on src/net.cpp:
```diff
@@ -2814,7 +2814,15 @@ bool CConnman::DisconnectNode(const CSubNet& subnet)
 bool CConnman::DisconnectNode(const CNetAddr& addr)
 {
-    return DisconnectNode(CSubNet(addr));
+    bool disconnected = false;
```
Is this change still needed with #20852? It seems to me that makes the old construction work?

vasild replied:
No, if #20852 is merged, then this PR is not needed, see https://github.com/bitcoin/bitcoin/pull/20849#issuecomment-754122796.

I first opened this PR to fix `DisconnectNode()` and then realized that the same problem exists in `BanMan`, thus opened #20852 which fixes both by modifying `CSubNet`.

Maybe this PR could be preferred as less risky compared to #20852. It is ok if this is merged first and later #20852 is also merged - they do not conflict and both go in `master`.

vasild commented:
Just to elaborate on the relationship between https://github.com/bitcoin/bitcoin/pull/20849 and https://github.com/bitcoin/bitcoin/pull/20852:

The problem in `master` is that `CSubNet(foo.onion)` creates an invalid subnet which later does not match any address, not even `foo.onion`. In two places in the code we expect that once we create such subnet later it will match `foo.onion`: in `CConnman::DisconnectNode()` and in `BanMan`.

https://github.com/bitcoin/bitcoin/pull/20849 fixes `DisconnectNode()` by changing it to not use subnets at all (small isolated change, easy to verify it is correct).

https://github.com/bitcoin/bitcoin/pull/20852 changes `CSubNet` to follow the above expected behavior, thus fixing both `DisconnectNode()` and `BanMan` (bigger change, harder to verify it is correct, caused unexpected CI failure (fixed)).

**About the release, maybe the following makes sense:**

Include https://github.com/bitcoin/bitcoin/pull/20849 in the next rc and in the `0.21.0` release.

Later, if https://github.com/bitcoin/bitcoin/pull/20852 is merged (ever) there is no need to revert https://github.com/bitcoin/bitcoin/pull/20849.

MarcoFalke commented:
It might be better to wait for #20852 (0.21.1), after which this is not needed.

MarcoFalke commented on src/test/denialofservice_tests.cpp:
```diff
@@ -223,47 +223,84 @@ BOOST_AUTO_TEST_CASE(peer_discouragement)
     const CChainParams& chainparams = Params();
     auto banman = MakeUnique<BanMan>(GetDataDir() / "banlist.dat", nullptr, DEFAULT_MISBEHAVING_BANTIME);
-    auto connman = MakeUnique<CConnman>(0x1337, 0x1337);
+    auto connman = MakeUnique<CConnmanTest>(0x1337, 0x1337);
```
std::make_unique for new code?

Also, could split the refactoring changes from the new-test changes?

MarcoFalke replied:
Nvm, forget the std::make_unique suggestion. We want to backport the test changes.

maflcko reviewed (commented):
Concept ACK on adding the test. Mind to open a new pull for this?

vasild commented:
The test is included in https://github.com/bitcoin/bitcoin/pull/20904.

taurus228 commented:
Join your team on Slack

taurussssssssss ***@***.***) has invited you to use Slack with them, in a workspace called Hi.

Workspace name: Hi

H

hi-vh78894.slack.com []

Join Now [https://join.slack.com/t/hi-vh78894/invite/enQtMjM3OTAyNTAyNTUwOS1jYjE2NGFlMzhlZTcwMzUwMzhjYTBhZTJjNWQ0NjY1MWQwNjhiMjc2YjkzZWU4OWMyNDQ3N2MzZTRhNTc5ODJh?x=x-p932605326480-921137271987-2367320120839]

taurussssssssss and 15 others have already joined

[Profile picture of taurussssssssss.][Profile picture of support.][Profile picture of walletfeedback-external.][Profile picture of taurussssss.][Profile picture of portis.][Profile picture of reply.][Profile picture of taurussssss2019.][Profile picture of community.]

What is Slack?

Slack is a messaging app for teams, a place you can collaborate on projects and organize conversations — so you can work together, no matter where you are. Learn more about Slack [https://hi-vh78894.slack.com/x-p932605326480-921137271987-2367320120839/?utm_medium=email&utm_source=invitation-to-join]
